test/e2e/targets_api_test.go

```diff
@@ -0,0 +1,144 @@
+		s.SharedDir(),
+		netName,
+		"prom2",
+		defaultPromConfig("ha", 1, "", ""),
+		e2ethanos.DefaultPrometheusImage(),
+	)
+	testutil.Ok(t, err)
+	testutil.Ok(t, s.StartAndWaitReady(prom1, sidecar1, prom2, sidecar2))
+
+	q, err := e2ethanos.NewQuerier(
+		s.SharedDir(),
+		"query",
+		[]string{sidecar1.GRPCNetworkEndpoint(), sidecar2.GRPCNetworkEndpoint()},
+		nil,
+		nil,
+		[]string{sidecar1.GRPCNetworkEndpoint(), sidecar2.GRPCNetworkEndpoint()},
+		"",
+		"",
+	)
+
+	testutil.Ok(t, err)
+	testutil.Ok(t, s.StartAndWaitReady(q))
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	t.Cleanup(cancel)
+
+	testutil.Ok(t, q.WaitSumMetricsWithOptions(e2e.Equals(2), []string{"thanos_store_nodes_grpc_connections"}, e2e.WaitMissingMetrics))
+
+	targetAndAssert(t, ctx, q.HTTPEndpoint(), "", &targetspb.TargetDiscovery{
+		ActiveTargets: []*targetspb.ActiveTarget{
+			{
+				DiscoveredLabels: labelpb.ZLabelSet{Labels: []labelpb.ZLabel{
+					{Name: "__address__", Value: "localhost:9090"},
+					{Name: "__metrics_path__", Value: "/metrics"},
+					{Name: "__scheme__", Value: "http"},
+					{Name: "job", Value: "myself"},
+					{Name: "prometheus", Value: "ha"},
+				}},
+				Labels: labelpb.ZLabelSet{Labels: []labelpb.ZLabel{
+					{Name: "instance", Value: "localhost:9090"},
+					{Name: "job", Value: "myself"},
+					{Name: "prometheus", Value: "ha"},
+				}},
+				ScrapePool: "myself",
+				ScrapeUrl:  "http://localhost:9090/metrics",
+				Health:     targetspb.TargetHealth_UP,
+			},
+		},
+		DroppedTargets: []*targetspb.DroppedTarget{
+			{
+				DiscoveredLabels: labelpb.ZLabelSet{Labels: []labelpb.ZLabel{
+					{Name: "__address__", Value: "localhost:80"},
+					{Name: "__metrics_path__", Value: "/metrics"},
+					{Name: "__scheme__", Value: "http"},
+					{Name: "job", Value: "myself"},
+					{Name: "prometheus", Value: "ha"},
+				}},
+			},
+		},
+	})
+}
+
+func targetAndAssert(t *testing.T, ctx context.Context, addr string, state string, want *targetspb.TargetDiscovery) {
+	t.Helper()
+
+	fmt.Println("targetAndAssert: Waiting for results for targets state", state)
+
+	testutil.Ok(t, runutil.Retry(time.Second, ctx.Done(), func() error {
+		res, err := promclient.NewDefaultClient().TargetsInGRPC(ctx, mustURLParse(t, "http://"+addr), state)
+		if err != nil {
+			return err
+		}
+
+		if len(res.ActiveTargets) != len(want.ActiveTargets) {
+			return errors.Errorf("unexpected result.ActiveTargets size, want %d; got: %d result: %v", len(want.ActiveTargets), len(res.ActiveTargets), res)
+		}
+
+		if len(res.DroppedTargets) != len(want.DroppedTargets) {
+			return errors.Errorf("unexpected result.DroppedTargets size, want %d; got: %d result: %v", len(want.DroppedTargets), len(res.DroppedTargets), res)
+		}
+
+		for it := range res.ActiveTargets {
+			res.ActiveTargets[it].LastScrape = time.Time{}
+			res.ActiveTargets[it].LastScrapeDuration = 0
+		}
+
+		for it := range res.ActiveTargets {
```

Comment:
You probably want here to iterate over `DroppedTargets`, right? But it seems like it doesn't matter since these have default values either way because they have never been scraped.